kani-compiler/src/kani_middle/transform/check_uninit/delayed_ub/instrumentation_visitor.rs

```diff
@@ -0,0 +1,125 @@
+    current: SourceInstruction,
+    /// The target instruction that should be verified.
+    pub target: Option<InitRelevantInstruction>,
+    /// Aliasing analysis data.
+    points_to: &'a PointsToGraph<'tcx>,
+    /// The list of places we should be looking for, ignoring others
+    analysis_targets: &'a HashSet<GlobalMemLoc<'tcx>>,
+    current_def_id: InternalDefId,
+    tcx: TyCtxt<'tcx>,
+}
+
+impl<'a, 'tcx> TargetFinder for InstrumentationVisitor<'a, 'tcx> {
+    fn find_next(
+        &mut self,
+        body: &MutableBody,
+        bb: BasicBlockIdx,
+        skip_first: bool,
+    ) -> Option<InitRelevantInstruction> {
+        self.skip_next = skip_first;
+        self.current = SourceInstruction::Statement { idx: 0, bb };
+        self.target = None;
+        self.visit_basic_block(&body.blocks()[bb]);
+        self.target.clone()
+    }
+}
+
+impl<'a, 'tcx> InstrumentationVisitor<'a, 'tcx> {
+    pub fn new(
+        points_to: &'a PointsToGraph<'tcx>,
+        analysis_targets: &'a HashSet<GlobalMemLoc<'tcx>>,
+        current_def_id: InternalDefId,
+        tcx: TyCtxt<'tcx>,
+    ) -> Self {
+        Self {
+            skip_next: false,
+            current: SourceInstruction::Statement { idx: 0, bb: 0 },
+            target: None,
+            points_to,
+            analysis_targets,
+            current_def_id,
+            tcx,
+        }
+    }
+    fn push_target(&mut self, source_op: MemoryInitOp) {
+        let target = self.target.get_or_insert_with(|| InitRelevantInstruction {
+            source: self.current,
+            after_instruction: vec![],
+            before_instruction: vec![],
+        });
+        target.push_operation(source_op);
+    }
+}
+
+impl<'a, 'tcx> MirVisitor for InstrumentationVisitor<'a, 'tcx> {
+    fn visit_statement(&mut self, stmt: &Statement, location: Location) {
+        if self.skip_next {
+            self.skip_next = false;
+        } else if self.target.is_none() {
+            // Check all inner places.
+            self.super_statement(stmt, location);
+            // Switch to the next statement.
+            let SourceInstruction::Statement { idx, bb } = self.current else { unreachable!() };
+            self.current = SourceInstruction::Statement { idx: idx + 1, bb };
+        }
+    }
+
+    fn visit_terminator(&mut self, term: &Terminator, location: Location) {
+        if !(self.skip_next || self.target.is_some()) {
+            self.current = SourceInstruction::Terminator { bb: self.current.bb() };
+            self.super_terminator(term, location);
+        }
+    }
+
+    fn visit_place(&mut self, place: &Place, ptx: PlaceContext, location: Location) {
+        // Match the place by whatever it is pointing to and find an intersection with the targets.
+        if self
+            .points_to
+            .follow_from_place(rustc_internal::internal(self.tcx, place), self.current_def_id)
+            .intersection(&self.analysis_targets)
+            .count()
```

Comment:
It's usually more efficient to use something like `.peekable().peek().is_some()` to check if it is non-empty (to avoid iterating over all elements).

Reply:
Thanks! I think that in this specific case we could even substitute it by `.next().is_some()`.

Reply:
Yes, since the iterator is not used for anything else, that should be fine.